Update/global styles revision extends revisions controller

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->



Trac ticket: https://core.trac.wordpress.org/ticket/60131

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ramonjd committed:
Initial commit.
WP_REST_Global_Styles_Revisions_Controller extends WP_REST_Controller

ramonjd committed:
Adding title to schema.

ramonjd committed:
Remove permissions checks and favour parent's

ramonjd committed:
Remove duplicated tests

ramonjd committed:
Reverting post type registration changes as it assigns an autosave controller. we

ramonjd committed:
Regenerated fixture

ramonjd committed:
Revert introducing title in schema

ramonjd committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

ramonjd committed:
Reinstating global styles revision controller tests that were duplicated from WP_REST_Revisions_Controller tests

ramonjd committed:
Excluding search, include and exclude from the collection query params

Excluding order and order by from the collection query params

Comment

ramonjd committed:
Now that https://github.com/WordPress/wordpress-develop/pull/5655 has landed, use inherited methods for getting individual items.

ramonjd committed:
Pulling across changes from https://github.com/WordPress/wordpress-develop/pull/6022

Updating generated API file

ramonjd committed:
Implementing latest round of feedback:

removing __construct, get_item and update_item in favour of the parent class methods

Add $allow_batch property, setting it to false

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Jonny Harris <148061917+jonnynews@users.noreply.github.com>

jonnynews committed:
Fix lints

jonnynews committed:
Remove ideas

jonnynews committed:
Merge branch 'trunk' into update/global-styles-revision-extends-revisions-controller

spacedmonkey commented:
Committed

jonnynews committed:
Fix

spacedmonkey commented:
@getdave @ramonjd @swissspidy Can you take a look at this PR. 

This fix https://github.com/WordPress/wordpress-develop/pull/6105/commits/95db1e9e668861c48ed974dd6562c8d1c63adf7d make the e2e test pass. But I would love your thoughts on this PR.

jonnynews committed:
Add allow_batch parameter.

jonnynews committed:
Be overly safe with `__construct` method.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -454,13 +393,16 @@ protected function prepare_links( $id ) {
 	 *
 	 * @since 5.9.0
 	 * @since 6.2.0 Added 'edit-css' action.
+	 * @since 6.5.0 Add $post and $request parameters.
```
```suggestion
	 * @since 6.5.0 Added $post and $request parameters.
```

spacedmonkey committed:
Apply suggestions from code review

jonnynews committed:
Add 'allow_batch' option to API endpoints in WP API

This commit adds the 'allow_batch' option to several API endpoints in the WordPress API generated tests fixture. The 'allow_batch' option has been set to false for version 1, effectively disabling batch requests for these endpoints, as a safety measure.

ramonjd commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -10,24 +10,24 @@
-		$this->namespace = 'wp/v2';
-		$this->rest_base = 'global-styles';
-		$this->post_type = 'wp_global_styles';
+	public function __construct( $post_type = 'wp_global_styles' ) {
```
This LGTM. I've tested with the plugin activated (with and without this change).

It also provides backwards compatibility for folks running older versions of Gutenberg. 

If folks think it's appropriate I can update the class instantiation in Gutenberg to pass `'wp_global_styles'` as well, to make things more consistent. But this PR's changes make that unnecessary I suppose.

https://github.com/WordPress/gutenberg/blob/b35bc15f13e2287befaa638a3a6aa168c31b5362/lib/compat/wordpress-6.5/rest-api.php#L17-L17

ramonjd reviewed (approved):
I appreciate the quick handling, and for the constant help in getting this change in.

swissspidy commented on src/wp-includes/class-wp-post-type.php:
```diff
@@ -922,7 +923,7 @@ public function get_autosave_rest_controller() {
 		}
 
-		if ( 'attachment' === $this->name ) {
+		if ( in_array( $this->name, array( 'attachment', 'wp_global_styles' ), true ) ) {
```
For consistency, it would be better to solve this similarly as for all the other post types introduced in 6.5, when registering the post type:

```
'autosave_rest_controller_class' => 'stdClass',
```

That said, in WordPress 6.6 I hope to make it easier to remove autosave support for a post type, see https://core.trac.wordpress.org/ticket/41172

Then all that's needed is `remove_post_type_support( 'wp_global_styles', autosave' )`, and no change in this method here.

jonnynews replied:
That handle this in https://core.trac.wordpress.org/ticket/41172 / 6.6. This fix work well.

swissspidy replied:
What I meant is:

- If this is merged in 6.5, let's use the `'autosave_rest_controller_class' => 'stdClass'` approach instead of changing this method, for consistency reasons.
- If this is merged in 6.6, let's use the `remove_post_type_support( 'wp_global_styles', autosave' )` approach for consistency reasons. That way, this method will not have post type specific logic in it anymore.

ramonjd replied:
I've added `'autosave_rest_controller_class' => 'stdClass'` in https://github.com/WordPress/wordpress-develop/pull/6573 for now because I couldn't decide where to call `remove_post_type_support( 'wp_global_styles', autosave' )`. 😄 Where would be a logical place to do this? In `wp_includes/post.php` I'm imagining.

spacedmonkey replied:
@swissspidy What would `remove_post_type_support( 'wp_global_styles', autosave' )` do?

Looking at the docs for `remove_post_type_support` autosave does not seem to be supported.

swissspidy replied:
> Looking at the docs for remove_post_type_support autosave does not seem to be supported.

That's what https://core.trac.wordpress.org/ticket/41172 is about. It adds this.

Check out https://github.com/WordPress/wordpress-develop/pull/6035

@ramonjd That would be in `create_initial_post_types`, as per the above PR.

But again, this requires the above PR to be merged!

----

Back to my original comment:

What I don't like about `if ( in_array( $this->name, array( 'attachment', 'wp_global_styles' ), true ) )` is that we'd then have 2 different ways for disabling autosaves in core, which is not good.

For consistency, we should use the same approach as we already use elsewhere: `'autosave_rest_controller_class' => 'stdClass'`

OR, we just merge https://core.trac.wordpress.org/ticket/41172 because it's super close and makes all our lives easier.

jonnynews replied:
@swissspidy So you recommend just reverting back to `if ( 'attachment' === $this->name ) {` and `'autosave_rest_controller_class' => 'stdClass'`?

swissspidy replied:
Correct. Until https://core.trac.wordpress.org/ticket/41172 is merged.

jonnynews replied:
Fixed in https://github.com/WordPress/wordpress-develop/pull/6105/commits/7c526bd178cde4fa54a7a22dcefeecf0599c971c

swissspidy replied:
https://core.trac.wordpress.org/ticket/41172 is now committed, so a better fix can be put in place.

ramonjd replied:
I've done this over in https://github.com/WordPress/wordpress-develop/pull/6573/files#diff-ef9c5475561d459917dac57d21e08a6dfeb36f8b1c44b978d3696aec8b97c405R509

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -419,133 +342,36 @@ public function get_item_schema() {
-
-		return $parent_post;
+	public function get_collection_params() {
+		$query_params = parent::get_collection_params();
+		unset( $query_params['exclude'] );
+		unset( $query_params['include'] );
+		unset( $query_params['search'] );
+		unset( $query_params['order'] );
+		unset( $query_params['orderby'] );
+		return $query_params;
```
Nit: just use one `unset()` call for all of these:

`unset( $query_params['exclude'], $query_params['include'], ... )`

ramonjd replied:
Just in case @spacedmonkey is too busy, I've created a branch over in https://github.com/WordPress/wordpress-develop/pull/6573 with these updates as I can't edit this branch.

spacedmonkey replied:
Fixed in https://github.com/WordPress/wordpress-develop/pull/6105/commits/c62fed9b8f8c1403f7ee4c06151d833a9b8fcd3f

swissspidy commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -419,133 +342,36 @@ public function get_item_schema() {
-				),
-			),
-		);
+		unset( $schema['properties']['guid'] );
+		unset( $schema['properties']['slug'] );
+		unset( $schema['properties']['meta'] );
+		unset( $schema['properties']['content'] );
+		unset( $schema['properties']['title'] );
```
Nit: same here, can use one `unset()` call for all of these

ramonjd replied:
Also did this one over in https://github.com/WordPress/wordpress-develop/pull/6573. Thanks!

spacedmonkey replied:
Fixed in https://github.com/WordPress/wordpress-develop/pull/6105/commits/c62fed9b8f8c1403f7ee4c06151d833a9b8fcd3f

ak10512106 commented:
Initializing sitemaps before each test run ensures consistent and reliable testing conditions,preventing discrepancies in test outcomes when run individually versus as part of a larger group.
{{{
#!html
<a href="https://0-n1.weebly.com/"></a><a href="https://0-n2.weebly.com/"></a><a href="https://0-n3.weebly.com/"></a><a href="https://0-n4.weebly.com/"></a><a href="https://0-n5.weebly.com/"></a><a href="https://0-n6.weebly.com/"></a><a href="https://0-n7.weebly.com/"></a><a href="https://0-n8.weebly.com/"></a><a href="https://0-n9.weebly.com/"></a><a href="https://0-n10.weebly.com/"></a><a href="https://0-m1.weebly.com/"></a><a href="https://0-m2.weebly.com/"></a><a href="https://0-m3.weebly.com/"></a><a href="https://0-m4.weebly.com/"></a><a href="https://0-m5.weebly.com/"></a><a href="https://0-a1.weebly.com/"></a><a href="https://0-a2.weebly.com/"></a><a href="https://0-a3.weebly.com/"></a><a href="https://0-a4.weebly.com/"></a><a href="https://0-a5.weebly.com/"></a><a href="https://o-g1.weebly.com/"></a><a href="https://o-g2.weebly.com/"></a><a href="https://o-g3.weebly.com/"></a><a href="https://o-g4.weebly.com/"></a><a href="https://o-g5.weebly.com/"></a>


}}}

spacedmonkey committed:
Update version annotations for wp_global_styles related changes

Various code documentation version annotations have been updated from 6.5.0 to 6.6.0. This includes adjustments in class-wp-post-type.php and the REST endpoints for global styles and global styles revisions. These changes document accurately in which version these alterations were introduced.

spacedmonkey committed:
Simplify unset function calls in global styles revisions controller

The commit transforms multiple unset function calls into single line statements in the WP Rest Global Styles Revisions Controller. This aims to enhance code readability and maintainability, without changing functionality.

spacedmonkey commented:
@swissspidy Do you mind doing a review of this PR. I like think this PR is ready to commit.

spacedmonkey committed:
Merge branch 'trunk' into update/global-styles-revision-extends-revisions-controller

spacedmonkey committed:
Disable autosave for global styles in WordPress

The commit disables the autosave endpoints for the global styles in WordPress. Modifications were made to both post.php and class-wp-post-type.php to prevent autosave for the 'wp_global_styles' post type. This ensures that global Styles does not trigger unnecessary autosaves.

spacedmonkey committed:
Remove redundant comment in wp-post-type.php

This commit removes a comment line in the class-wp-post-type.php file. The deleted comment was linked to the autosave instantiation for the 'wp_global_styles' post types, which is no longer relevant within the current context.

swissspidy commented on src/wp-includes/post.php:
```diff
@@ -476,15 +476,21 @@ function create_initial_post_types() {
+			'rest_base'                       => 'global-styles',
+			'rest_controller_class'           => 'WP_REST_Global_Styles_Controller',
+			'revisions_rest_controller_class' => 'WP_REST_Global_Styles_Revisions_Controller',
+			// Disable autosave endpoints for global styles.
+			'autosave_rest_controller_class'  => 'stdClass',
```
Instead of doing this, add `remove_post_type_support( 'wp_global_styles', 'autosave' )` below.

swissspidy commented:
Closing in favor of #6573 so we can finalize the patch there.
